argilla/src/argilla/records/_dataset_records.py

```diff
@@ -358,25 +360,35 @@ def to_datasets(self) -> HFDataset:
     def _ingest_records(
         self,
         records: Union[List[Dict[str, Any]], Dict[str, Any], List[Record], Record, HFDataset],
-        mapping: Optional[Dict[str, str]] = None,
+        mapping: Optional[Dict[str, Union[str, List[str], Tuple[str]]]] = None,
         user_id: Optional[UUID] = None,
     ) -> List[RecordModel]:
+        """Ingests records from a list of dictionaries, a Hugging Face Dataset, or a list of Record objects."""
+
         if len(records) == 0:
             raise ValueError("No records provided to ingest.")
+
         if HFDatasetsIO._is_hf_dataset(dataset=records):
             records = HFDatasetsIO._record_dicts_from_datasets(dataset=records)
-        if all(map(lambda r: isinstance(r, dict), records)):
-            # Records as flat dicts of values to be matched to questions as suggestion or response
-            records = [self._infer_record_from_mapping(data=r, mapping=mapping, user_id=user_id) for r in records]  # type: ignore
-        elif all(map(lambda r: isinstance(r, Record), records)):
-            for record in records:
-                record.dataset = self.__dataset
-        else:
-            raise ValueError(
-                "Records should be a a list Record instances, "
-                "a Hugging Face Dataset, or a list of dictionaries representing the records."
-            )
-        return [record.api_model() for record in records]
+
+        ingested_records = []
+        for record in tqdm(records, desc="1/2: Ingesting records", unit="record"):
+            try:
+                if not isinstance(record, Record):
+                    rendered_mapping = self._render_record_mapping(records=records, mapping=mapping)
+                    record = self._infer_record_from_mapping(data=record, mapping=rendered_mapping, user_id=user_id)
```

Comment:
I think we should merge both methods in once
```suggestion
                    record = self._infer_record_from_mapping(data=record, mapping=mapping, user_id=user_id)                    
```
and, internally, the `_infer_record_from_mapping` will have a call to                     rendered_mapping = self._render_record_mapping(records=records, mapping=mapping)

Reply:
I've updated this method so that we only call `_render_record_mapping` once before the loop. Do you think we should call it every time?